packages/framework-core/src/services/raw-events-parser.ts

```diff
@@ -1,14 +1,46 @@
 /* eslint-disable @typescript-eslint/no-explicit-any */
-import { BoosterConfig, EventEnvelope } from '@boostercloud/framework-types'
+import { BoosterConfig, EventEnvelope, Logger, UUID } from '@boostercloud/framework-types'
+
+export type EventsStreamingCallback = (
+  entityName: string,
+  entityID: UUID,
+  eventEnvelopes: Array<EventEnvelope>,
+  config: BoosterConfig
+) => Promise<void>
+type EnvelopesPerEntity = Record<string, Array<EventEnvelope>>
 
 export class RawEventsParser {
-  public static async streamEvents(
+  public static async streamPerEntityEvents(
+    logger: Logger,
     config: BoosterConfig,
     rawEvents: any,
-    callbackFn: (eventEnvelope: EventEnvelope, config: BoosterConfig) => Promise<void>
+    callbackFn: EventsStreamingCallback
   ): Promise<void> {
-    for (const event of config.provider.events.rawToEnvelopes(rawEvents)) {
-      await callbackFn(event, config)
+    const eventEnvelopesPerEntity = config.provider.events
+      .rawToEnvelopes(rawEvents)
+      .filter(isEventKind)
+      .reduce(groupByEntity, {})
```

Comment:
Now we process events in batches, not one by one (this was very inefficient before). So we need to remove snapshots and divide the events by entity ID, so they are safely reduced together